vector-im/element-android
# FTUE - Homeserver sign in/up deeplinks

Opened as a draft as this relies on #5995 

## Type of change


## Content

- Adds missing deeplink handling to the new FTUE combined login/registration screens
- Fixes offline errors when attempting to sign in being treated as an invalid homeserver selection
  - Introduces a helper for inferring if the device has an active connection, if not we can further breakdown unknown host errors and infer that the failure is caused by being offline rather than the homeserver being unavailable 
- Lifts the invalid homeserver error dialog to the Activity(/FtueVariant) with a retry action passed from the `ViewModel`

## Motivation and context

Fixes #6023 

## Screenshots / GIFs

#### Legacy flows

| WITH FTUE DISABLED | GET STARTED WITH FTUE DISABLED | LEGACY RESETTING DEEPLINK |
| --- | --- | --- |
![after-deeplink-legacy-sign-in](https://user-images.githubusercontent.com/1848238/168087798-6490acaa-c400-459c-aa3c-0bbbc8e0383b.gif)|![after-deeplink-legacy-get-started](https://user-images.githubusercontent.com/1848238/168087803-a0d17d32-a98e-46da-b29b-b77e38dffa62.gif)|![after-deeplink-legacy-error](https://user-images.githubusercontent.com/1848238/168087807-2ef19b0f-a486-4174-9199-9ef7be076f90.gif)|

#### FTUE flows

| SIGN UP | SIGN IN |
| --- | --- |
![after-deeplink-sign-up-error](https://user-images.githubusercontent.com/1848238/168087819-0c2ab698-e033-423b-aefc-aa4512af872a.gif)|![after-deeplink-new-sign-in-error](https://user-images.githubusercontent.com/1848238/168087812-a69b6d09-1a23-4a34-b829-403b8dfef679.gif)

| SIGN UP | SIGN IN | 
| --- | --- |
|![deeplink-reg-disabled](https://user-images.githubusercontent.com/1848238/168091628-dd453a02-91a5-4336-8686-bae49bd40568.gif)|![after-sign-in-valid](https://user-images.githubusercontent.com/1848238/168091621-18c3ce78-8699-4a44-8ab4-51e51a99d873.gif)

#### Offline deeplink

|BEFORE | AFTER |
| --- | --- |
|![Screenshot_20220512_144406](https://user-images.githubusercontent.com/1848238/168089461-71b9cded-3633-4cde-bf76-2e187526eb4e.png)|![after-offline-sign-in](https://user-images.githubusercontent.com/1848238/168088324-2eaa3b84-6d4b-4c9b-b76b-0f57ffbb4ad6.png)


## Tests

- Open the app via valid and invalid deeplinks 
  - Invalid - https://mobile.element.io/?hs_url=matrix.example.com&is_url=identity.example.com 
  - Valid with registration disabled - https://mobile.element.io/?hs_url=iswell.cool&is_url=identity.example.com 
- Sign in or sign up
- Notice homeserver is preselected or an error to choose a different homeserver is shown 

## Tested devices

- OS version(s): 28
